[Enhancement] make LIKE operations' statistics consistent

ahmadov opened the pull request:
## Why I'm doing:

The `LIKE(column, pattern)` function and the `column LIKE pattern` operator have different statistics in
[ExpressionStatisticCalculator](https://github.com/StarRocks/starrocks/blob/41843adc63d8bb9c968370dc4d97acfb1b1a0685/fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/ExpressionStatisticCalculator.java).

This leads to an issue when it's used together with other functions (e.g. `IF`).

Example:
`IF(column LIKE pattern, 'Yes', 'No')` estimates 2 distinct values as expected, however when the same query is executed again as `IF(LIKE(column, pattern), 'Yes', 'No')` the distinct value estimation is now different than 2 because of this check [[1]].

## What I'm doing:
Since they are the same operation, their statistics should be same too. The `LikePredicateOperator` does not have a direct estimation in `ExpressionStatisticCalculator`, instead it takes the statistics of the first child by default [[2]].
With this change, the same applies to the `LIKE` function and it takes the statistics of the first child (left hand side).

[1]: https://github.com/StarRocks/starrocks/blob/41843adc63d8bb9c968370dc4d97acfb1b1a0685/fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/ExpressionStatisticCalculator.java#L189
[2]: https://github.com/StarRocks/starrocks/blob/41843adc63d8bb9c968370dc4d97acfb1b1a0685/fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/ExpressionStatisticCalculator.java#L74

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.4
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0

ahmadov committed:
[Enhancement] make LIKE operations' statistics consistent

The `LIKE(column, pattern)` function and the `column LIKE pattern`
operator have different statistics in
[ExpressionStatisticCalculator](https://github.com/StarRocks/starrocks/blob/41843adc63d8bb9c968370dc4d97acfb1b1a0685/fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/ExpressionStatisticCalculator.java).

This leads to an issue when it's used together with other functions (e.g. `IF`).

Example:
`IF(column LIKE pattern, 'Yes', 'No')` estimates 2 distinct values as
expected, however when the same query is executed again as
`IF(LIKE(column, pattern), 'Yes', 'No')` the distinct value estimation
is now different than 2 because of this check [[1]].

Since they are the same operation, their statistics should be same too.
The `LikePredicateOperator` does not have a direct estimation in
`ExpressionStatisticCalculator`, instead it takes the statistics of the
first child by default [[2]].
With this change, the same applies to the `LIKE` function and it takes
the statistics of the first child (left hand side).

[1]: https://github.com/StarRocks/starrocks/blob/41843adc63d8bb9c968370dc4d97acfb1b1a0685/fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/ExpressionStatisticCalculator.java#L189
[2]: https://github.com/StarRocks/starrocks/blob/41843adc63d8bb9c968370dc4d97acfb1b1a0685/fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/ExpressionStatisticCalculator.java#L74

Signed-off-by: Elmi Ahmadov <e.ahmadov@celonis.com>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/StarRocks/starrocks?pullRequest=55023) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/StarRocks/starrocks?pullRequest=55023) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/StarRocks/starrocks?pullRequest=55023) it.</sub>

ahmadov committed:
Adjust min/max/distinct count for the LIKE function

The LIKE function returns boolean as a result, therefore the values
should be adjusted according to that behaviour.

Signed-off-by: Elmi Ahmadov <e.ahmadov@celonis.com>

ahmadov committed:
Add tests to improve coverage for both `LIKE` and `ILIKE` functions

Signed-off-by: Elmi Ahmadov <e.ahmadov@celonis.com>

ahmadov commented on fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/ExpressionStatisticCalculator.java:
```diff
@@ -613,6 +613,12 @@ private ColumnStatistic binaryExpressionCalculate(CallOperator callOperator, Col
                     distinctValues = left.getDistinctValuesCount() * averageRowSize / left.getAverageRowSize();
                     break;
+                case FunctionSet.LIKE:
+                    minValue = left.getMinValue();
```
update the PR. According to [LIKE docs](https://docs.starrocks.io/docs/sql-reference/sql-functions/like-predicate-functions/like/), output values are 0 (false) and 1 (true), therefore I used the same values for min and max values.
